### PR TITLE
Refactoring, fixes and support for extra-component-references-label

### PR DESCRIPTION
**Pull request by 8R0WNI3**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```other operator

```

### CODE REVIEW

File: cnudie/iter.py [hunk 1]
```diff
@@ -1,22 +1,41 @@
...
 
-import gci.componentmodel as cm
 import cnudie.retrieve
+import dso.labels
+import gci.componentmodel as cm
+
+
+class NodeReferenceType(enum.StrEnum):
+    COMPONENT_REFERENCE = 'componentReference'
+    EXTRA_COMPONENT_REFS_LABEL = dso.labels.ExtraComponentReferencesLabel.name
```
Comment (ccwienk) [inline cnudie/iter.py:12]: should we prefix the label (e.g. with `label:`)?

File: cnudie/iter.py [hunk 2]
```diff
@@ -1,22 +1,41 @@
...
     def component(self):
-        return self.path[-1]
+        return self.path[-1].component
 
     @property
     def component_id(self):
         return self.component.identity()
 
+    def __post_init__(self):
+        self.path = tuple(
```
Comment (ccwienk) [inline cnudie/iter.py:34]: shouldn't we rather expect callers to pass correct values (we might check + raise ValueError if values are incorrect)?

File: cnudie/iter.py [hunk 3]
```diff
@@ -31,14 +50,16 @@ class ArtefactNode:
     - iterable (useful for pattern-matching, e.g. c,a = node)
     '''
     @property
-    def artefact(self) -> cm.Resource | cm.ComponentSource:
+    def artefact(self) -> cm.Resource | cm.Source:
```
Comment (ccwienk) [inline cnudie/iter.py:53]: good catch

File: cnudie/iter.py [hunk 4]
```diff
@@ -140,17 +162,43 @@ def inner_iter(
...
+                referenced_component_descriptor = lookup(extra_cref_id, ocm_repo)
+            else:
+                referenced_component_descriptor = lookup(extra_cref_id)
+
+            yield from inner_iter(
+                component=referenced_component_descriptor.component,
                 lookup=lookup,
                 recursion_depth=recursion_depth,
                 path=path,
+                reftype=NodeReferenceType.EXTRA_COMPONENT_REFS_LABEL,
```
Comment (ccwienk) [inline cnudie/iter.py:201]: IMHO, we should only mark the exact component that declared reference via label (sub-components from such a component should again be marked as regular crefs (if they are regular crefs, of course)

**Review by ccwienk — CHANGES_REQUESTED** [review]

File: cnudie/iter.py (hunk 4 above)
Comment (8R0WNI3) [reply cnudie/iter.py:201]: I was not sure about this either, so I'm with you to define it like that :-)

File: ctt/uploaders.py [hunk 5]
```diff
@@ -149,7 +149,7 @@ def process(self, processing_job, target_as_source=False):
 
         src_prefix, src_name, src_tag = oci.client._split_image_reference(src_ref)
 
-        if ':' in src_tag:
+        if '@' in src_tag:
```
Comment (ccwienk) [inline ctt/uploaders.py:152]: preferably use oci.model.OciImageReference + `has_tag`

File: cnudie/iter.py [hunk 6]
```diff
@@ -105,9 +107,12 @@ def iter(
                             dependencies; -1 will resolve w/o recursion limit, 0 will not resolve
                             component dependencies
     @param prune_unique: if true, redundant component-versions will only be traversed once
-    @node_filter:        use to filter emitted nodes (see Filter for predefined filters)
+    @param node_filter:  use to filter emitted nodes (see Filter for predefined filters)
     @param ocm_repo:     optional OCM Repository to be used to override in the lookup
     @param ctx_repo:     deprecated, use `ocm_repo` instead
+    @param component_filter: use to exclude components (and their references) from the iterator
```
Comment (ccwienk) [inline cnudie/iter.py:113]: it would be nice to state that returning `True` means component should be filtered out

File: cnudie/iter.py [hunk 7]
```diff
@@ -105,9 +107,12 @@ def iter(
...
+    @param reftype_filter: use to exclude components (and their references) from the iterator if
```
Comment (ccwienk) [inline cnudie/iter.py:114]: ditto

File: ctt/process_dependencies.py [hunk 8]
```diff
@@ -420,12 +424,20 @@ def process_images(
 
     executor = concurrent.futures.ThreadPoolExecutor(max_workers=16)
 
+    reftype_filter = None
+    if skip_labels and skip_labels(dso.labels.ExtraComponentReferencesLabel.name):
+        def filter_extra_component_refs(reftype: cnudie.iter.NodeReferenceType) -> bool:
+            return reftype == cnudie.iter.NodeReferenceType.EXTRA_COMPONENT_REFS_LABEL
```
Comment (ccwienk) [inline ctt/process_dependencies.py:430]: nit: use `is` for comparison w/ enum-values?

did you use `==` to gracefully handle cases where we encounter a str-value we do not know as of our enum?

Comment (8R0WNI3) [reply ctt/process_dependencies.py:430]: No good point, I did not do this on purpose. The `reftype` here stems from `cnudie.iter`, so we can assume that it's a value of our enum or at least fail if it's not 👍🏼

File: ctt/process_dependencies.py [hunk 9]
```diff
@@ -688,6 +690,8 @@ def reupload_component(component: cm.Component):
 
         if processing_mode is ProcessingMode.REGULAR:
             if component.name == source_comp.name and component.version == source_comp.version:
+                if skip_root_cd_upload:
```
[the remark this answers is not in the document]
Comment (8R0WNI3) [reply ctt/process_dependencies.py]: I also thought about it again. I came out with the solution to keep the `skip_component_upload` parameter with its current semantics (preventing just the cd-replication, not the replication of its resources and component references). Instead, I added a newly introduced parameter `component_filter` which has the semantics of excluding the resources and transitive component references from the replication as well. I also added some comments explaining this behaviour 😅

File: dso/labels.py [hunk 10]
```diff
@@ -91,6 +91,18 @@ class CveCategorisationLabel(Label):
...
 
+@dataclasses.dataclass(frozen=True)
+class ExtraComponentReference:
+    component_reference: cm.ComponentIdentity
+    purpose: tuple[str] | None
+
+
+@dataclasses.dataclass(frozen=True)
+class ExtraComponentReferencesLabel(Label):
+    name = 'ocm.software/ocm-gear/extra-component-references'
```
Comment (zkdev) [inline dso/labels.py:102]: Shouldn't this rather be a general "defined / well-known" ocm label, and not an `ocm-gear` / `delivery-gear` specific?

File: ctt/process_dependencies.py [hunk 11]
```diff
@@ -648,6 +661,15 @@ def append_ctx_repo(ctx_base_url: str | cm.OciOcmRepository, component: cm.Compo
         ocm_repository = component.current_repository_ctx()
         oci_ref = ocm_repository.component_version_oci_ref(component)
 
+        if skip_labels:
+            labels: list[cm.Label] = []
+
+            for label in component.labels:
+                if not skip_labels(label.name):
+                    labels.append(label)
```
Comment (zkdev) [inline ctt/process_dependencies.py:669]: rather use list comprehension 🙂

**Review by zkdev — CHANGES_REQUESTED** [review]

**Review by zkdev — APPROVED** [review]
/lgtm

**Pull request merged**